### ISSUE TITLE
Does the Distributed Data Parallel (DDP) require additional GPU memory for maintaining a model?

### ISSUE THREAD

**Issue by xiuguangLi** [pytorch/pytorch#131653]
Does the Distributed Data Parallel (DDP) require additional GPU memory for maintaining a model?

Here is the code:

python
```
print(f"before 1 {torch.cuda.memory_reserved()/(1024**2)}")  ---> 0
model = model.to(self.device)
print(f"before 2 {torch.cuda.memory_reserved()/(1024**2)}") ---> 172
model = torch.nn.parallel.DistributedDataParallel(
                        model, 
                        device_ids=device_ids, output_device=output_device, **kwargs
                    )
print(f"after ddp {torch.cuda.memory_reserved()/(1024**2)}") ---> 324
model.to("cpu")
torch.cuda.empty_cache()
print(f"after cpu {torch.cuda.memory_reserved()/(1024**2)}")  ---> 158
```
In this case, it can be observed that the model size occupies 166, and the CUDA overhead occupies 6. The remaining 158-6=152 is the additional cost generated by DDP. Is there a way to remove or move this additional overhead to the CPU? Because I have many models to train, every time I wrap them with DDP, it will generate additional overhead, causing my code to run out of GPU memory.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by H-Huang** [comment]
cc @fegin do you know why wrapping with DDP would increase memory usage?

**Comment by wconstab** [comment]
My guess is, because DDP is playing a trick on you in order to optimize communications.  It wants to do one large allreduce for a whole bunch of parameters, but that requires the parameters share one contiguous buffer.  When the model was first initialized, its parameters would not share one contiguous space.  So probably DDP allocates a new contiguous space large enough for the parameters inside one bucket, then replaces the original model parameters with views into that bucket, hence  freeing the original storage for those parameters.  

I'm not familiar with how DDP is actually doing this, or whether it does so bucket-by-bucket or allocates memory for all buckets at once before freeing any original parameters.

Is this causing issues for you to fit a model into the GPU?

**Comment by fegin** [comment]
@xiuguangLi The additional reserved memory is from DDP reducer. DDP reducer pre-allocate gradient memory which will be later used for gradient reduction. Note that reducer will always allocate memory during the initialization. But you can ask DDP to reuse the memory space for gradients. Please check `gradient_as_bucket_view` option of DDP, which will use this memory for gradients. We won't be able to get rid of this extra memory because during training you will need to allocate memory for gradients.
